### PR TITLE
Add function to validate membership proofs for quotes

### PR DESCRIPTION
**Pull request by wjuan-mob**
This PR is intended to add validation that the quotes added to the deqs are valid quotes by checking that the mixin inputs are real. 
It also includes a change to the scibuilder test utils to add an optional ledger argument to enable tests to add the Txos they want to select as mixin inputs to the ledger so that the synchronized quotebook won't reject the quotes for not being valid.

### CODE REVIEW

**Comment by cbeck88** [comment]
I think we should rethink this:

* Quotes that come in to deqs should not have attached membership proof, because those are very large
* Instead we just verify by looking at the ledger that each mixin input is real, by looking up based on its global tx out index, and equality testing the result

File: quote-book/synchronized/src/lib.rs [hunk 1]
```diff
@@ -76,6 +76,31 @@ impl<Q: QuoteBook, L: Ledger + Clone + Sync + 'static> SynchronizedQuoteBook<Q,
...
         self.highest_processed_block_index.load(Ordering::SeqCst)
     }
+
+    fn validate_quote_ring_members(&self, quote: &Quote) -> Result<(), Error> {
+        let indices = &quote.sci().tx_out_global_indices;
+        let ring = &quote.sci().tx_in.ring;
+        //We should have an index per txo.
+        if ring.len() != indices.len() {
+            return Err(Error::ImplementationSpecific(
+                "Missing index for ring".to_owned(),
```
Comment (cbeck88) [inline quote-book/synchronized/src/lib.rs:86]: maybe we can add dedicated error types for this? "InvalidRingMember" or something?

**Comment by cbeck88** [comment]
this looks generally good to me with some tests, thank you

File: quote-book/api/src/error.rs [hunk 2]
```diff
@@ -26,6 +26,9 @@ pub enum Error {
     /// Quote cannot fulfill the desired amount ({0}) of base tokens
     InsufficientBaseTokens(u64),
 
+    /// Quote has an invalid ring
+    InvalidRing(String),
```
Comment (eranrund) [inline quote-book/api/src/error.rs:30]: This should get added to `api/src/convert/quote_book_error.rs`

File: test-cli/src/bin/main.rs [hunk 3]
```diff
@@ -123,6 +123,7 @@ fn main() {
                             0,
                             counter_amount,
                             &mut rng,
+                            None,
```
Comment (eranrund) [inline test-cli/src/bin/main.rs:126]: Passing `None` here means the test utility no longer generates SCIs that can get accepted by the server. I think a potential for solution for this would be for the test-cli to accept a ledger file and append to it similar to what the tests are doing. That unfortunately does mean that all servers have to share the same ledger file when running this locally. It's acceptable for now, since I think we could use the liquidity bot I am working on for submitting SCIs out of real accounts.
2)

**Review by eranrund — CHANGES_REQUESTED** [review]
Nice, thank you. Mostly LGTM but it breaks the test cli that I use to manually run servers on my local machine, so that needs to be addressed.

File: mc-test-utils/src/lib.rs [hunk 4]
```diff
@@ -22,19 +23,35 @@ pub fn create_sci_builder(
...
         &offerer_account,
         &fog_resolver,
         rng,
     );
+    if let Some(mut ledger) = ledger_db {
+        let num_txos = ledger.num_txos().unwrap();
+        offered_input_credentials
+            .membership_proofs
+            .iter_mut()
+            .zip(0u64..)
```
[the remark this answers is not in the document]
Comment (wjuan-mob) [reply mc-test-utils/src/lib.rs:45]: It didn't work with enumerate because it gives a usize. The error message suggested zipping with 0u64..

[the remark this answers is not in the document]
Comment (eranrund) [reply mc-test-utils/src/lib.rs:45]: I see. Okay. I personally find it a bit counterintuitive and would've casted the usize to u64 but it doesnt matter that much

**Review by eranrund — CHANGES_REQUESTED** [review]
Unfortunately this is broken. When I try to run it (`./target/release/deqs-test-cli --ledger-db ../internal/mobilecoin/sample_data/ledger submit-quotes  --deqs-uri insecure-deqs://127.0.0.1/  --num-quotes 50`) I get this:
```
...
2023-03-06 19:34:11.632694 UTC INFO Generating 500 SCIs..., mc.app: deqs-test-cli, mc.module: deqs_test_cli, mc.src: test-cli/src/bin/main.rs:107
thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: Invalid block index: 487', /Users/eran/Projects/mc/deqs/mc-test-utils/src/lib.rs:53:10
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: Invalid block index: 487', /Users/eran/Projects/mc/deqs/mc-test-utils/src/lib.rs:53:10
thread 'thread 'thread '<unnamed><unnamed><unnamed>' panicked at '' panicked at '' panicked at 'called `Result::unwrap()` on an `Err` value: Invalid block index: 487called `Result::unwrap()` on an `Err` value: Invalid block index: 487called `Result::unwrap()` on an `Err` value: Invalid block index: 487', ', /Users/eran/Projects/mc/deqs/mc-test-utils/src/lib.rs:53:thread '10<unnamed>
' panicked at 'thread '<unnamed>', ' panicked at '/Users/eran/Projects/mc/deqs/mc-test-utils/src/lib.rscalled `Result::unwrap()` on an `Err` value: Invalid block index: 487:', 53thread 'called `Result::unwrap()` on an `Err` value: Invalid block index: 487/Users/eran/Projects/mc/deqs/mc-test-utils/src/lib.rs<unnamed>', /Users/eran/Projects/mc/deqs/mc-test-utils/src/lib.rs:/Users/eran/Projects/mc/deqs/mc-test-utils/src/lib.rs' panicked at ':53:called `Result::unwrap()` on an `Err` value: Invalid block index: 48710:10
:53:53:
10
', /Users/eran/Projects/mc/deqs/mc-test-utils/src/lib.rs:53:1010

thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: Invalid block index: 487', /Users/eran/Projects/mc/deqs/mc-test-utils/src/lib.rs:53:10
thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: Invalid block index: 488', /Users/eran/Projects/mc/deqs/mc-test-utils/src/lib.rs:53:10
thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: Invalid block index: 489', /Users/eran/Projects/mc/deqs/mc-test-utils/src/lib.rs:53:10
thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: Invalid block index: 489', /Users/eran/Projects/mc/deqs/mc-test-utils/src/lib.rs:53:10
thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: Invalid block index: 489', /Users/eran/Projects/mc/deqs/mc-test-utils/src/lib.rs:53:10
thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: Invalid block index: 489', /Users/eran/Projects/mc/deqs/mc-test-utils/src/lib.rs:53:10
...
```
How was this tested?

It makes sense due to the `.into_par_iter()` call.
The easy solution is to change it to `.iter()` but that makes things slower. Probably acceptable for the time being.

Additionally, please resolve merge conflicts before marking a PR ready for review. Thanks!

File: test-cli/src/bin/main.rs [hunk 5]
```diff
@@ -67,6 +67,10 @@ pub struct Config {
 
     #[clap(subcommand)]
     pub command: Command,
+
+    /// Path to ledgerdb used by the server
+    #[clap(long, env = "MC_LEDGER_DB")]
+    pub ledger_db: PathBuf,
```
Comment (eranrund) [inline test-cli/src/bin/main.rs:73]: This should be under the Subcommand, it is not needed for all of the commands.

File: test-cli/src/bin/main.rs [hunk 6]
```diff
@@ -98,21 +105,12 @@ fn main() {
...
-                .collect::<Vec<_>>();
-
-            let scis = rng_seeds
-                .into_par_iter()
-                .map(|rng_seed| {
-                    let mut rng: StdRng = SeedableRng::from_seed(rng_seed);
+                    // generate a unique seed for each SCI we will be generating.
+                    let mut rng_seed = [0; 32];
+                    rng.fill_bytes(&mut rng_seed);
+                    let mut reseeded_rng: StdRng = SeedableRng::from_seed(rng_seed);
```
Comment (eranrund) [inline test-cli/src/bin/main.rs:113]: Why is this needed? The idea of the seed is that all SCIs end up generated deterministically and I believe that would happen if `rng` is used, without this stuff. Am I missing something?

File: quote-book/test-suite/src/lib.rs [hunk 7]
```diff
@@ -30,13 +31,15 @@ pub fn create_sci_builder(
     base_amount: u64,
     counter_amount: u64,
     rng: &mut (impl RngCore + CryptoRng),
+    ledger_db: Option<LedgerDB>,
```
Comment (eranrund) [inline quote-book/test-suite/src/lib.rs:34]: Nit: I suggest switching to `Option<&LedgerDB>` in all these create_ methods, and do the cloning inside. It will save typing a whole bunch of `.clone()` calls everywhere.

**Review by eranrund — CHANGES_REQUESTED** [review]

File: test-cli/src/bin/main.rs [hunk 8]
```diff
@@ -47,6 +47,10 @@ pub enum Command {
         /// Allow partial fills
         #[clap(long)]
         allow_partial_fills: bool,
+
+        /// Path to ledgerdb used by the server
+        #[clap(long, env = "MC_LEDGER_DB")]
+        ledger_db_path: PathBuf,
```
Comment (eranrund) [inline test-cli/src/bin/main.rs:53]: The server and other parts of the codebase use `--ledger-db`, we should be consistent. 
You can use `long = "ledger-db"`, or just change the var name to `ledger_db`, it's fine if on line 104 the variable gets aliased (I think the compiler would be okay with that)

**Review by eranrund — APPROVED** [review]
